_drafts/drafting-a-c-development-environment.md

```diff
@@ -0,0 +1,810 @@
+
+```bash
+$ make builder-build
+$ make builder-run
+
+$ clang-format --version
+Debian clang-format version 16.0.6 (++20230704084212+7cbf1a259152-1~exp1~20230704204302.101)
+
+$ clang-tidy --version
+Debian LLVM version 16.0.6
+  Optimized build.
+```
+
+### Using `clang-format` in the Development Container
+
+`vscode` typically needs to be instructed to rebuild the development container from the _Dockerfile_. You can do so explicitly using the command prompt:
+
+![]({% img_url c-dev-environment/vscode-rebuild.png %})
+
+Now you have `clang-tidy` and `clang-format` available within your development container. You can now also instruct `vscode` to automatically format your files on save. For this, we create a workspace configuration [`.vscode/settings.json`](https://github.com/lmapii/cproject/blob/main/.vscode/settings.json) file:
+
+```json
+{
+  "[c]": {
+    "editor.formatOnSave": true
+  },
+  "[cpp]": {
+    "editor.formatOnSave": true
+  },
+  "C_Cpp.default.compileCommands": "${workspaceFolder}/src/build/compile_commands.json",
+  "cmake.configureOnOpen": true
+}
+```
+
+I've hidden a little detail here: `clang-tidy` operates on the `compile_commands.json` file that can be created when using `cmake` as a build system. I've enabled this option within the [`CMakeLists.txt`](https://github.com/lmapii/cproject/blob/main/CMakeLists.txt) file. At the same time, we're allowing the `cmake` extension for `vscode` to configure the project when opening and therefore have all the fancy buttons and actions available to build the project in case you don't like the command line. Personally, I still fall back to the command line. But then, I'm a dinosaur.
+
+Now, if you open the `dummy.c` file in your editor it will automatically format it according to the rules in your `.clang-format` file whenever you save the file. But what if someone in your team doesn't use `vscode` or forgot to enable the format-on-save feature?
+
+### Wrapping `clang-format` and `clang-tidy` calls
+
+This section is a little self-promotion of two wrapper scripts for `clang-format` and `clang-tidy` that I am maintaining: [run-clang-tidy](https://github.com/lmapii/run-clang-tidy) and [run-clang-format](https://github.com/lmapii/run-clang-format). Both tools are command line tools written in [Rust](https://www.rust-lang.org/) that simplify and parallelize the execution of `clang-format` and `clang-tidy` from the command line.
+
+Installing the tools within your container _should_ be easy via the [Rust package manager `cargo`](https://github.com/rust-lang/cargo), however, at the time of writing, I could not install any binaries inside a container via the `cargo` `apt` package, and installing the tools from scratch within the container simply took too much time, so I decided to download the released, architecture-specific version instead:
```

Comment:
That's surprising. Does using the rustup download script not work either?

Reply:
So, installing using `rustup` seems to work but in increases the image size by 1 GB and compiling both tools from scratch takes another 10 minutes on my machine. I extended the article to mention this - this is a common pro/con evaluation when installing tools in a Dockerfile.